distribution/lib/Standard/Base/0.0.0-dev/src/Data/Statistics.enso

```diff
@@ -279,3 +319,46 @@ Vector.compute self statistic=Statistic.Count =
 Vector.compute_bulk : Vector Statistic -> Vector Any
 Vector.compute_bulk self statistics=[Statistic.Count, Statistic.Sum] =
     Statistic.compute_bulk self statistics
+
+## Compute a single running statistic on the vector.
+
+   Arguments:
+   - statistic: Statistic to calculate.
+Vector.running : Statistic -> Vector Any
+Vector.running self statistic=Statistic.Count =
+    Statistic.running self statistic
+
+## PRIVATE
+compute_fold c v =
+    if is_valid v . not then c else
+         c.first.increment
+
+         if c.first.comparatorError.not then
+             if c.first.minimum.is_nothing then c.first.setMinimum v else
+                 ordering = Incomparable_Values.handle_errors <| v.compare_to c.first.minimum
+                 if ordering.is_error then c.first.failComparator else
+                     if ordering == Ordering.Less then c.first.setMinimum v
+             if c.first.maximum.is_nothing then c.first.setMaximum v else
+                 ordering = Incomparable_Values.handle_errors <| v.compare_to c.first.maximum
+                 if ordering.is_error then c.first.failComparator else
+                     if ordering == Ordering.Greater then c.first.setMaximum v
+
+         new_moments = if c.second.is_nothing then Nothing else case v of
+             _ : Number -> c.second.add v
+             _ -> Nothing
+
+         [c.first, new_moments]
```

Comment:
If we always pass `c.first`, I don't see any point of making it an argument of `fold`.

If the counter was an immutable structure we would pass its new values as part of the accumulator. But the counter here is completely imperative, so passing it in an accumulator is IMO misleading - it creates a false feeling that we are doing a functional fold where really we rely on an imperative structure under the hood anyway. This confused me and made it harder to decipher what this code does.

As the counter reference stays the same for the whole iteration I'd suggest adding it as a separate parameter to this function - then we can give it a meaningful name too.

```diff
-compute_fold c v =
+compute_fold counter acc value =
 ...
-[c.first, new_moments]
+new_moments   
```

Then at the callsite we could do:
```diff
-result = compute_fold [counter, ref_moment.get] v
-ref_moment.put result.second
+new_moment = compute_fold counter ref_moment.get v
+ref_moment.put new_moment
```

Reply:
At the other callsite we can do:
```diff
-result = data.fold [counter, moments] compute_fold
+result = data.fold moments (compute_fold counter)
```

IMO this is clearer.